packages/controller/src/tasks/syncTenants.ts

```diff
@@ -21,7 +21,58 @@ import { KubeConfig } from "@kubernetes/client-node";
 import { set as updateTenants, Tenants } from "@opstrace/tenants";
 import dbClient, { ClientResponse } from "../dbClient";
 import { State } from "../reducer";
-import { log, SECOND } from "@opstrace/utils";
+import { log, SECOND, httpcl } from "@opstrace/utils";
+import { Response as GotResponse } from "got";
+
+function* setDefaultAlertmanagerConfigIfEmpty(tenant: string) {
+  const res: GotResponse<string> = yield httpcl(
+    `http://alertmanager.cortex.svc.cluster.local/api/v1/alerts`,
+    {
+      method: "GET",
+      headers: {
+        "X-Scope-OrgID": tenant
+      },
+      throwHttpErrors: false
+    }
+  );
+
+  if (res.statusCode > 499) {
+    // Network error, don't potentially overwrite existing config
+    log.error(
+      `could not read alertmanager config for tenant: ${tenant}, got resp: ${res.statusCode}, ${res.statusMessage}`
+    );
+    return;
+  }
+
+  if (res.body.length) {
```

Comment:
if the response is a 404 Not Found one that this might be a false positive -- why not explicitly check for a 2xx response?

Reply:
The multi-tenant Alertmanager in Cortex responds with a 404 for the endpoint if it hasn't been configured yet, I.e. if it hasn't received a POST with a valid config before. 😞

Reply:
I am not sure if I got my point across, and I am also not sure if I understand your reply.

Trying my point again: this code block contains the comment `// Already has config set`. Hence, I am assuming that it really is meant to handle a success response, indicating that a config is set for this tenant. Subsequently I assume that this is always wrapped in a 2xx response, probably a 200 one. Why do we not explicitly check for that expected status code?

The criterion implemented right now is `res.body.length` being non-zero. This is fulfilled for any HTTP response having a non-empty body. This is usually the case for all error responses.

As of the `if (res.statusCode > 499)` above, this condition is met for _all responses_ with a non-empty body and a status code `<= 499`, including most 4xx responses. This still seems wrong to me :). If it is correct, it kind of needs better explanation.

That is why I assume that this condition is met for cases where `// Already has config set` is _not_ true.

Reply:
Maybe we want something like this:

```
if (res.statusCode === 200) {
    // definite signal: a config for this tenant has been set
    return;
}

if (res.statusCode !== 404 {
  // Handle 5xx responses: ambiguous result, we do not know if
  // a config has already been set for this tenant or not. Back out.   
  // Also handle all 4xx responses except for 404; they are unexpected
  // at this point and should be logged properly.
  log.error(...)
  return
}

// When we are here, we know the response had status code 404.
// definite signal that no config for this tenant has been set yet
...do the POST request...

Reply:
PR with suggested changes: https://github.com/opstrace/opstrace/pull/902